[5.4] Add ability to use Action Buttons with Slack Notifications Channel

aaronhuisinga opened the pull request:
Slack has recently added the ability to use Action Buttons when sending messages via Slack Apps. These allow users to perform programmed actions by simply clicking an Action Button contained in a message.

This PR adds support for using these buttons to the Slack Notification Channel. It's a simple extension of the already existing attachment method for Slack Notifications.

The PR is fully backward compatible, as it introduces no breaking changes - just new methods for adding these Action Buttons. 

@taylorotwell If this PR is merged, I'd be happy to put together a second PR for adding documentation to the Laravel website!

Slack documentation for Action Buttons: https://api.slack.com/docs/message-buttons
Slack Action Buttons "Field Guide": https://api.slack.com/docs/interactive-message-field-guide

A demo of the new method with action buttons and a confirmation window for a destructive action:
```php
public function toSlack($notifiable)
{
    return (new SlackMessage)
        ->content('Content')
        ->attachment(function ($attachment) {
            $attachment->title('Laravel', 'https://laravel.com')
                ->content('Attachment Content')
                ->callbackId('laravel_123')
                ->action(function ($action) {
                   $action->name('laravel')
                          ->text('Cancel')
                          ->style('danger')
                          ->type('button')
                          ->value('laravel-cancel')
                          ->confirmation(function($confirm) {
                              $confirm->title('Are You Sure?')
                                      ->content('Are you sure you want to cancel?')
                                      ->okText('Yes, Cancel')
                                      ->dismissText('Dismiss');
                          });
                })
                ->action(function ($action) {
                    $action->name('laravel')
                           ->text('Okay')
                           ->type('button')
                           ->value('laravel-okay');
                });
        });
}
```

And a demo of the method using a select dropdown for providing action options:
```php
public function toSlack($notifiable)
{
    return (new SlackMessage)
        ->content('Content')
        ->attachment(function ($attachment) {
            $attachment->title('Laravel', 'https://laravel.com')
                ->content('Attachment Content')
                ->callbackId('laravel_123')
                ->action(function ($action) {
                   $action->name('laravel')
                          ->text('Cancel')
                          ->style('danger')
                          ->type('select')
                          ->value('laravel-cancel')
                          ->option(function($option) {
                            $option->text('First Option')
                                    ->value('first_option');
                          })
                          ->option(function($option) {
                              $option->text('Second Option')
                                     ->value('second_option');
                          });
                })
                ->action(function ($action) {
                    $action->name('laravel')
                           ->text('Okay')
                           ->type('button')
                           ->value('laravel-okay');
                });
        });
}
```

Commit:
Add ability to use Slack Action Buttons with Notifications

Commit:
Add callback_id field to Slack attachments

Commit:
Add test to cover Slack Action buttons

Commit:
StyleCI fixes

Commit:
One more StyleCI fix

jakebathman commented:
This is a really cool idea, nice job.

lucasmichot commented on src/Illuminate/Notifications/Messages/SlackAttachmentAction.php:
```diff
@@ -0,0 +1,255 @@
+     *
+     * @return $this
+     */
+    public function confirmationFields(array $confirmationFields)
```
What if we want to pass an object that implements `\ArrayAccess` ?

aaronhuisinga replied:
I did this due to several other of the related Slack notification methods type-hinting to array. I can definitely remove that if that's the preference. No huge deal to me.

lucasmichot commented on src/Illuminate/Notifications/Channels/SlackWebhookChannel.php:
```diff
@@ -111,4 +114,21 @@ protected function fields(SlackAttachment $attachment)
+     */
+    protected function actions(SlackAttachment $attachment)
+    {
+        return collect($attachment->actions)->map(function ($value) {
```
Do we really need a `Collection`?

This can be achieved with:
```php
return array_map(function ($action) {
    return $action instanceof SlackAttachmentAction ? $action->toArray() : $action;
}, (array) $attachment->actions);
```

aaronhuisinga replied:
Makes sense to me 👍 I'd just gone along with formatting similar to some related Slack Notification methods. This seems simpler and is likely faster. Thanks @lucasmichot!

aaronhuisinga replied:
@taylorotwell @themsaid I can definitely switch this to an array map instead of using a collection. I just modeled it after a similar method.

lucasmichot commented on src/Illuminate/Notifications/Messages/SlackAttachment.php:
```diff
@@ -175,6 +189,43 @@ public function fields(array $fields)
+     */
+    public function action($title, $content = [])
+    {
+        if (is_callable($title)) {
```
Isn't it faster to just do:
```php
if (is_callable($title)) {
    $this->actions[] = $title(new SlackAttachmentAction);
} else {
    $this->actions[$title] = $content;
}

return $this;
```

aaronhuisinga replied:
I modeled this after the field method on line 166. Does it seem to make sense to refactor both of these? I had a similar thought but decided to model after a method already in the repo.

taylorotwell commented:
There is a ton of stuff here that is not in your tests... minQueryLength... etc I have no idea what that stuff is. I've fixed a lot of formatting issues on my end already by hand so if you could just explain in the comment what all of these other things that aren't tested do that would be helpful.

aaronhuisinga commented:
@taylorotwell Sure thing. I just cloned the parameters that are available for action buttons from the Slack API docs. I'll go ahead and update the comments in my methods to be more detailed.

Slack's Action Button parameter docs, for reference: https://api.slack.com/docs/interactive-message-field-guide#message_action_fields

Commit:
Update docblocks for new classes to be more detailed for methods

taylorotwell commented:
I meant there are a lot of methods in this that aren't in your code example in your comment or in your tests.

aaronhuisinga commented:
Whoops - sorry about that. I'll get them both updated this morning!

Commit:
Fix several bugs with Slack Actions and improve tests to cover all methods

Commit:
StyleCI fixes

Commit:
One more StyleCI fix

aaronhuisinga commented:
@taylorotwell I've updated the examples in the PR description to include two separate examples that should cover all available methods. I've also added more methods to the existing test and added a second test to make sure everything is all covered. We should be good to go here.

taylorotwell commented:
Going to hold off on this for now. It's just so many methods to keep up with and so wordy. My intention with notifications was to build notifications, not really build these interactive buttons and such. Will revisit later maybe.

aaronhuisinga commented:
Ah - bummer. I'll just refactor this to a custom notification method that we can use for this purpose. Thanks anyway, @taylorotwell

guerinj commented:
@aaronhuisinga Did you refactor this in something easily shareable ? (package, gist, ...) 

Thanks a lot :)

fitztrev commented:
Basic link buttons were just added via laravel/slack-notification-channel#6.

However, you still can't do **[Interactive Buttons](https://api.slack.com/docs/message-buttons)** which this PR implements. Example of an Interactive Button:
![](https://a.slack-edge.com/dcb1/img/api/message_guidelines/Example_6.gif)

I would still really like to see that functionality. Maybe now that slack-notification-channel is its own repo and that other PR has been merged, Taylor will reconsider.

aaronhuisinga commented:
@fitztrev We've been using a version of my PR here in production since it was closed here. I just wrote a custom Channel class that extended the necessary framework Slack classes. This PR should still work properly without any changes, I believe. If you want to open a PR in the new Slack Notification Channel repo, definitely go for it. Just give me a shoutout somewhere there so I can see if it gets merged and remove my implementation 👍
